Docs/doc 95/doc 102/technical term action

Rachel-Reverie opened the pull request:
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- Adds technical term page for Action (previously Validation Action)



After submitting your PR, CI checks will run and @ge-cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).


### Definition of Done
Please delete options that are not relevant.

- [X] I have made corresponding changes to the documentation


Thank you for submitting!

Rachel-Reverie committed:
Updated Validation Action to Action in the files and terms.json
Added Action content and page

Rachel-Reverie committed:
Added tags to technical terms.

Rachel-Reverie committed:
Corrected keys in some technical tags.

Rachel-Reverie committed:
- updated information on `data_asset` parameter in Actions to indicate it will be a Validator instance.

joshua-stauffer commented on docs/terms/action.md:
```diff
@@ -0,0 +1,109 @@
+
+### Definition
+
+An Action is a Python class with a `run()` method that takes the result of validating a <TechnicalTag relative="../" tag="batch" text="Batch" /> against an <TechnicalTag relative="../" tag="expectation_suite" text="Expectation Suite" /> and does something with it.
```
open question - would this be clearer/more concise to say `An Action is a Python class with a run() method that takes a Validation Result and does something with it`?

Rachel-Reverie replied:
I do think that is more concise.  I've updated the working definition.

joshua-stauffer commented on docs/terms/action.md:
```diff
@@ -0,0 +1,109 @@
+
+## Use cases
+
+Creating custom Actions is a process that falls outside the workflow of using Great Expectations.  Configuring, implementing, and executing an Action (custom or otherwise) takes place in the Validate Data step.
```
This first sentence took me a minute to put into context -- but after reading the second sentence I see it's relating to the main steps of GE. It is definitely okay as is, but i may have had less friction if the order of the sentences was switched.

Rachel-Reverie replied:
I've switched the order of these statements.

talagluck commented on docs/terms/action.md:
```diff
@@ -0,0 +1,109 @@
+
+### Relationship to other objects
+
+Actions are executed by <TechnicalTag relative="../" tag="validation_operator" text="Validation Operators" />, which are a component of Checkpoints.  In general, Validation Operators function behind the scenes, so it is enough to know that Actions are configured inside the `action_list` parameter of a Checkpoint's configuration, and execute every time the Checkpoint finishes running a Validation.
```
I would think that we may not want to even mention `ValidationOperators` since it could lead to some confusion. That said, if someone else feels differently, and that it is helpful to maintain the docs reference, I think the way you have it hear with the disclaimers and warnings is the right way to do it.

Rachel-Reverie replied:
I'd like to leave the reference in place purely for completeness and accuracy's sake -- I can imagine a scenario in which someone is working on extending Great Expectations in some way that requires knowledge of how Actions are executed.

talagluck commented on docs/terms/action.md:
```diff
@@ -0,0 +1,109 @@
+
+The required and optional named parameters will be automatically passed to the Action from the Checkpoint that the Action is included with, after the Checkpoint completes Validation.  This means you can configure your custom Actions to behave conditionally based on the Validation Results your Checkpoint generates, or the values passed along with any of those named parameters.  Additional `**kwarg ...
+
+The best practice when creating a custom Action is actually to create a subclass of the `ValidationAction` class found in the `great_expectations.checkpoint.actions` module.  Leave the inherited `run()` method as its default, and overwrite the `_run()` method to contain your functionality.  There are numerous examples of this practice in the subclasses of `ValidationAction` located in the `great_ ...
```
```suggestion
The best practice when creating a custom Action is to create a subclass of the `ValidationAction` class found in the `great_expectations.checkpoint.actions` module.  Leave the inherited `run()` method as its default, and overwrite the `_run()` method to contain your functionality.  There are numerous examples of this practice in the subclasses of `ValidationAction` located in the `great_expectations.checkpoint.actions` module, which you can view on GitHub:
```

talagluck commented on docs/terms/action.md:
```diff
@@ -0,0 +1,109 @@
+
+### Configuration
+
+Actions are configured inside of the `action_list` parameter for Checkpoints.  In general, you will need to provide at least a `name` (which is user defined and does not need to correspond to anything specific) and, in the in the parameters under `action` a `class_name` (which should correspond to the name of the Action's python class).  If you are implementing a custom Action in a plugin, you wi ...
```
```suggestion
Actions are configured inside of the `action_list` parameter for Checkpoints.  In general, you will need to provide at least a `name` (which is user defined and does not need to correspond to anything specific) and, in the in the parameters under `action` a `class_name` (which should correspond to the name of the Action's python class).  If you are implementing a custom Action in a Plugin, you will also need to include a `module_name` in the parameters under `action` which references where your Plugin is located..  Any other keys placed under `action` will be passed to the Action's class as additional key word arguments.
```

talagluck commented on docs/terms/action.md:
```diff
@@ -0,0 +1,109 @@
+
+Actions are configured inside of the `action_list` parameter for Checkpoints.  In general, you will need to provide at least a `name` (which is user defined and does not need to correspond to anything specific) and, in the in the parameters under `action` a `class_name` (which should correspond to the name of the Action's python class).  If you are implementing a custom Action in a plugin, you wi ...
+
+Below is an example of an `action_list` configuration that performs some common actions by indicating Actions that are built in to the Great Expectations code base:
```
```suggestion
Below is an example of an `action_list` configuration that performs some common Actions that are built in to the Great Expectations code base:
```

talagluck reviewed (approved):
LGTM overall! A few questions and minor updates (non-blocking). 

One additional question - would it be appropriate to put a note somewhere in the section on creating custom Actions encouraging a contribution if a user thinks that others might benefit?

Rachel-Reverie committed:
Update docs/terms/action.md

Co-authored-by: talagluck <tal@superconductive.com>

Rachel-Reverie committed:
Update docs/terms/action.md

Co-authored-by: talagluck <tal@superconductive.com>

Rachel-Reverie committed:
Merge branch 'universal-map-docs-update' into DOCS/DOC-95/DOC-102/technical_term_action

Rachel-Reverie committed:
Update docs/terms/action.md

Co-authored-by: talagluck <tal@superconductive.com>

Rachel-Reverie commented on docs/terms/action.md:
```diff
@@ -1,21 +1,108 @@
+
+### Configuration
+
+Actions are configured inside of the `action_list` parameter for Checkpoints.  In general, you will need to provide at least a `name` (which is user defined and does not need to correspond to anything specific) and, in the in the parameters under `action` a `class_name` (which should correspond to the name of the Action's python class).  If you are implementing a custom Action in a Plugin, you wi ...
```
```suggestion
Actions are configured inside of the `action_list` parameter for Checkpoints.  In general, you will need to provide at least a `name` (which is user defined and does not need to correspond to anything specific) and, in the in the parameters under `action` a `class_name` (which should correspond to the name of the Action's python class).  If you are implementing a custom Action in a Plugin, you will also need to include a `module_name` in the parameters under `action` which references where your Plugin is located.  Any other keys placed under `action` will be passed to the Action's class as additional key word arguments.
```

Rachel-Reverie committed:
Update docs/terms/action.md

joshua-stauffer commented on docs/terms/action.md:
```diff
@@ -1,21 +1,108 @@
+
+### How to access
+
+Actions are not intended to be manually instantiated or accessed.  Instead, they are included in the `action_list` parameter of a Checkpoint's configuration, and when the Checkpoint finishes running a Validation it will then run the Actions in it's `action_list`.
```
Might be useful context to add `... then run the Actions in its action_list in order of appearance`. (also, i think that apostrophe may have snuck in there by mistake)

Rachel-Reverie replied:
I've added this clarification.  And fixed the apostrophe.

joshua-stauffer commented on docs/terms/action.md:
```diff
@@ -1,21 +1,108 @@
+
+### How to create
+
+Custom actions need to be created in Python code, and can be implemented as Plugins.  In order for a python class to be a valid Action, it must conform to the Action API.  Specifically, it must implement a `run()` method which accepts three required parameters, two named optional parameters, and any number of `**kwargs.`
```
consistent casing for Python

Rachel-Reverie replied:
Corrected, and I double checked the document for any other missed instances.

joshua-stauffer reviewed (approved):
This is great, @Rachel-Reverie! I added a couple thoughts/suggestions, but feel free to use or disregard them.

Rachel-Reverie committed:
- Updated Action definition to be more concise
- Added note regarding contributions and Slack consultation in relation to developing custom Actions.
- Corrected casing for Python
- Added clarification regarding order Actions are run in.

Rachel-Reverie commented:
> LGTM overall! A few questions and minor updates (non-blocking).
> 
> One additional question - would it be appropriate to put a note somewhere in the section on creating custom Actions encouraging a contribution if a user thinks that others might benefit?

That would likely be appropriate.  I've checked with @kyleaton and added a note in regarding contributions and reaching out on Slack if assistance is needed.
